pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,474 @@
+}
+
+// Resolved is a resolved name.  Once the Value has been used successfully,
+// users MUST call `(*Resolved).Record(..)` to possibly record it as a new
+// short-name alias.
+type Resolved struct {
+	// Fully-qualified reference with tag or digest.
+	Value reference.Named
+	// Input from the user.
+	inputName string
+	// Control whether to record it permanently as an alias.
+	record bool
+	// Allows to reason over pull errors and add some context information.
+	rationale rationale
+	// Path to a config file where a given artifact is specified.
+	config string
+	// The system context used when looking up the alias.  Always recording
+	// the system context allows for easier testing as we can simulate
+	// recording even if `record` would inititally be set false.
+	systemContext *types.SystemContext
+}
+
+// Allows to reason over pull errors and add some context information.
+// Used in (*Resolved).WrapPullError.
+type rationale int
+
+const (
+	// No additional context.
+	rationaleNone rationale = iota
+	// Resolved value is a short-name alias.
+	rationaleAlias
+	// Resolved value has been completed with an Unqualified Search Registry.
+	rationaleUSR
+	// Resolved value has been selected by the user (via the prompt).
+	rationaleUserSelection
+)
+
+func newResolved(name string, value reference.Named, record bool, r rationale, config string, sys *types.SystemContext) Resolved {
+	return Resolved{
+		Value:         value,
+		inputName:     name,
+		record:        record,
+		rationale:     r,
+		config:        config,
+		systemContext: sys,
+	}
+}
+
+// Record stores the alias iff the alias has been set to be stored permanently
+// by `Resolve`.
+func (r *Resolved) Record() error {
+	if !r.record {
+		return nil
+	}
+	if err := Add(r.systemContext, r.inputName, r.Value.String()); err != nil {
+		return errors.Wrapf(err, "error recording short-name alias (%q=%q)", r.inputName, r.Value)
+	}
+	return nil
+}
+
+// WrapPullError wraps the specified error and adds some context information
+// such as if the `Resolved` originates from a short-name alias, an
+// unqualified-search registry or if it was selected by the user in a prompt.
+func (r *Resolved) WrapPullError(err error) error {
+	if err == nil {
+		return err
+	}
+
+	var conf string
+	if len(r.config) > 0 {
+		conf = fmt.Sprintf(" (see %s)", r.config)
+	}
+
+	switch r.rationale {
+	case rationaleAlias:
+		return errors.Wrapf(err, "%q is a recorded short-name alias%s", r.Value.String(), conf)
+	case rationaleUSR:
+		return errors.Wrapf(err, "%q has been completed with an unqualified-search registry%s", r.Value.String(), conf)
+	case rationaleUserSelection:
+		return errors.Wrapf(err, "%q has been selected by the user", r.Value.String())
+	case rationaleNone:
+		fallthrough
+	default:
+		return err
+	}
+}
+
+// FormatPullErrors is a convenience function to format errors that occured
+// while trying to pull one or more `Resolved` items.
+func FormatPullErrors(pullErrors []error) error {
```

Comment:
This is more or less an UI aesthetics decision; it _can_ live here but it’s not all that necessary (compare https://github.com/hashicorp/go-multierror as previously used in one of the pull implementations).

As it is, it does not easily allow implementing the de-duplication (it still requires the caller to individually `WrapError` every single one, and then here we would have to un-wrap that, and we have neither a `Resolved` nor any other object to carry state, so the full state of the resolution would have to be passed through every single `Resolved` and every single error returned through `WrapError` only to be deduplicated here.

And what happens if the caller mistakenly returns `pullErrors` from more than one `Resolve` call?

Basically it seems to me that we will deprecate this after the first improvement.

---

```go
type ResolutionResult {
   Results []reference.Named // or an array of structs? Not sure, we can always add a top-level array to expand.
   unqualified-search-origin string
   itemRationale []rationale
}

func (rr *ResolutionResult) FormatPullErrors(errors []error) string { // or WrapPullErrors that returns an error? Not sure; It's not very machine-consumable
  // errors must match Canonical
  // Just list them, or add context, or prioritize, or whatever
}

// Possible FUTURE extensions (let's not expand this PR, please)

// Returns a short-form description suitable for progress displays before starting to pull (e.g. without including paths and other info that is only useful for diagnosing failures)
func (rr *ResolutionResult) ShortDescription() string

// Returns a short-form description suitable for progress displays, before starting to pull rr.Results[index] (e.g. without including paths and other info that is only useful for diagnosing failures)
func (rr *ResolutionResult) ResultContext(index index) string

// Pull in order, give appropriate progress updates, and so on
func c/common/sharedpull.PullResolvedImage(rr *ResolutionResult, progressWriter, …)
```

Reply:
You're making a pretty convincing point. Encapsulating all data in a single type is a wise thing to do.